Basic Inlet docs (#326)
LLNL/axom: src/axom/inlet/docs/sphinx/quick_start.rst

.. _inlet_quick_start_label:

Quick Start
===========

Inlet's workflow is broken into the four following steps:  

* Defining the schema of your input deck

> * Verifying that the user given input deck is valid

Reviewer: Suggestion:
"user given" -> "user provided" (or maybe "user-provided")

> * Verifying that the user given input deck is valid

Reviewer: Suggestion for second step:  
> Read in and validate the user-provided input

Author: reading happens during the defining stage. It has been suggested to break reading and defining apart.

> * Accessing the read and stored data in your program

Reviewer: Suggestion for step 3:
> Accessing the input data in your program

* *Optional* Generating documentation based on defined schema

.. _inlet_defining_schema_label:

Defining Schema
---------------

The first step in using Inlet is to define the schema of your input deck.
Inlet defines an input deck into two basic classes: Tables and Fields. Basically
Fields are individual values and Tables hold groups of Fields and Tables.

Define the schema by using the following functions, on either the main Inlet class, for

> global Tables and Fields, or on individual Table classes, for Tables and Fields under that Table:

Reviewer: Suggestion: "individual Table classes" -> "individual Table instances" ?

========================= ===================
addTable                  Adds a Table to the input deck schema with the given name.
addBool                   Adds a boolean Field to the global or parent Table with the given name.
addDouble                 Adds a double Field to the global or parent Table with the given name.
addInt                    Adds a integer Field to the global or parent Table with the given name.
addString                 Adds a string Field to the global or parent Table with the given name.
========================= ===================

All possible Tables and Fields that are can be found in the input deck must be defined
at this step.  The value of the Field is read and stored into the Sidre datastore when you call the appropriate
add function. Use the ``required`` class member function on the Table and Field class to indicate that
they have to present in the given input deck. You can also set a default value to each field via the type-safe
``Field::defaultValue()`` member functions. Doing so will populate the corresponding Fields value
if the specific Field is not present in the input deck. The following example shows these concepts:

> .. literalinclude:: ../../examples/verification.cpp

Reviewer: Nice example!

   :start-after: _inlet_workflow_defining_schema_start
   :end-before: _inlet_workflow_defining_schema_end
   :language: C++

.. _inlet_verification_label:

Verification
------------

This step helps ensure that the given input deck follows the rules expected by the code.  These

> rules are not verified until you call ``Inlet::verify()``.  Doing so will return true/false and

Reviewer: Clarification: (I think it says this, but we might want to be explicit)
The `verify` happens after all data has been parsed, so the verification rules can access any other part of the input, right?

output SLIC warnings to indicate which Field or Table violated which rule.

As shown above, both Tables and Fields can be marked as ``required``. Fields have two additional
basic rules that can be enforced with the following ``Field`` class member functions:

========================= ===================
Name                      Description
========================= ===================
validValues               Indicates the Field can only be set to one of the given values.
range                     Indicates the Field can only be set to inclusively between two values.
========================= ===================

Inlet also provides functionality to write your own custom rules via callable lambda verifiers.
Fields and Tables can both register one lambda each via their ``registerVerifier()`` member functions.
The following example adds a custom verifier that simply verifies that the given ``dimensions`` field
match the length the given vector:

> .. literalinclude:: ../../examples/verification.cpp

Reviewer: ```cpp
  // We expect verification to be unsuccessful since the only Field
  // in vector is x but 2 dimensions are expected
  std::cout << "This should fail due to a missing dimension:\n";
  myInlet->verify() ? std::cout << "Verification was successful\n" 
                    : std::cout << "Verification was unsuccessful\n";
```

Should these use SLIC macros? Or are they intentionally using `cout`?

   :start-after: _inlet_workflow_verification_start
   :end-before: _inlet_workflow_verification_end
   :language: C++

.. note::  ``Inlet::getGlobalTable()->registerVerifier()`` can be used to add a verifier to apply rules
  to the Fields at the global level.

.. _inlet_accessing_data_label:

Accessing Data
--------------

After the input deck has been read and verified by the previous steps, you can access the data by name
via ``Inlet::get()`` functions.  These functions are type-safe, fill the given variable with what is found,
and return a boolean whether the Field was present in the input deck or had a default value it could fall
back on.  Variables, on the Inlet side, are used in a language-agnostic way and are then converted 
to the language-specific version inside of the appropriate ``Reader``. For example, Inlet refers to the
Lua variable ``vector={x=3}`` or ``vector.x`` as ``vector/x`` on all Inlet function calls.

For example, given the previous verificiation example, this access previously read values:

.. literalinclude:: ../../examples/verification.cpp
   :start-after: _inlet_workflow_accessing_data_start

>    :end-before: _inlet_workflow_accessing_data_end

Reviewer: Consider showing both `vector.x` and vector/x` notations in this example
(with comments to indicate that these are two different ways to access the data)?

e.g.
```cpp
  x_found = myInlet->get("vector/x", x);   // you can access element using '/' as a separator
  y_found = myInlet->get("vector.y", y);   // you can also use '.' as a separator
```

Author: you can't use vector.y at this point everything must be language agnostic

   :language: C++

.. _inlet_generating_documentation_label:

Generating Documentation
------------------------

We provide a slightly more complex but closer to a real world Inlet usage example of the usage of Inlet.
You can find that example in our repository `here <https://github.com/LLNL/axom/blob/develop/src/axom/inlet/examples/document_generation.cpp>`_.

After you create your ``Inlet`` class but before you start defining your schema, create a concrete
instantiation of a ``DocWriter`` class and register it with your ``Inlet`` class.

.. literalinclude:: ../../examples/documentation_generation.cpp
   :start-after: _inlet_documentation_generation_start
   :end-before: _inlet_documentation_generation_end
   :language: C++

Then after you are finishing defining your schema, call ``writeDoc()`` on your ``Inlet`` class
to write out your documentation to the given file.

.. code-block:: C++

>    inlet->writeDoc();

Reviewer: Would it make sense to show a screenshot of the output? Or an rst codeblock ?

Author: Adding the document_generation output. lets see how weird it looks...

.. _inlet_quick_start_label:

Quick Start
===========

Inlet's workflow is broken into the three following steps:  

* Defining the schema of your input deck
* Verifying that the user given input deck is valid
* Accessing the read and stored data in your program

.. _inlet_defining_schema_label:

Defining Schema
---------------

> The first step in using Inlet is to defining the schema of your input deck.

Reviewer 2: ```suggestion
The first step in using Inlet is to define the schema of your input deck.
```

they have to present in the given input deck. You can also set a default value to each field via the type-safe
``Field::defaultValue()`` member functions. Doing so will populate the corresponding Fields value
if the specific Field is not present in the input deck. The following example shows these concepts:

.. literalinclude:: ../../examples/verification.cpp
   :start-after: _inlet_workflow_defining_schema_start
   :end-before: _inlet_workflow_defining_schema_end
   :language: C++

.. _inlet_verification_label:

Verification
------------

This step helps ensure that the given input deck follows the rules expected by the code.  These
rules are not verified until you call ``Inlet::verify()``.  Doing so will return true/false and
output SLIC warnings to indicate which Field or Table violated which rule.

As shown above, both Tables and Fields can be marked as ``required``. Fields have two additional
basic rules that can be enforced with the following ``Field`` class member functions:

========================= ===================
Name                      Description
========================= ===================

> validValues               Indicates the Field can only be set to a valid values.

Reviewer 2: ```suggestion
validValues               Indicates the Field can only be set to a valid value.
```

range                     Indicates the Field can only be set to inclusively between two values.
========================= ===================

Inlet also provides functionality to write your own custom rules via callable lambda verifiers.
Fields and Tables can both register one lambda each via their ``registerVerifier()`` member functions.

> The following example adds a custom verifier that simply verifies that the dimension of the simulation

Reviewer 2: ```suggestion
The following example adds a custom verifier that simply verifies that the dimensions of the simulation
```

match up with the dimensions of a given vector:

.. literalinclude:: ../../examples/verification.cpp
   :start-after: _inlet_workflow_verification_start
   :end-before: _inlet_workflow_verification_end
   :language: C++

.. note::  ``Inlet::getGlobalTable()->registerVerifier()`` can be used to add a verifier to apply rules
  to the Fields at the global level.

.. _inlet_accessing_data_label:

Accessing Data
--------------

> After the input deck has been read and verified by the previous steps.  You can access the data by name

Reviewer 2: ```suggestion
After the input deck has been read and verified by the previous steps, you can access the data by name
```

via ``Inlet::get()`` functions.  These functions are type-safe, fill the given variable with what is found,
and return a boolean whether the Field was present in the input deck or had a default value it could fall

> back on.  Variable are named in a language agnostic way and are converted from Inlet's representation 

Reviewer 2: ```suggestion
back on.  Variables are named in a language-agnostic way and are converted to Inlet's representation 
```

> to the language specific version inside of the appropriate ``Reader``. For example, Inlet refers to the

Reviewer 2: ```suggestion
from the language-specific version inside of the appropriate ``Reader``. For example, Inlet refers to the
```
